### PR TITLE
design for flow crud operations refactor

### PR DESCRIPTION
**Pull request by timofei-durakov**
proposal is to move flow crud operations to
WFM flow topology

### CODE REVIEW

File: docs/design/solutions/hub-and-spoke-tpe/flow_create_install.txt [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+control kilda.northbound
+participant "FLOW SplitterBolt" as SplitterBolt
+participant "FLOW CrudBolt" as CrudBolt
+participant "FLOW ErrorBolt" as ErrorBolt
+participant "FLOW NorthboundReplyBolt" as NorthboundReplyBolt
+participant "FLOW FlowCommandBolt" as FlowCommandBolt
+participant FlowService
+participant DAO
+participant "FLOW TransactionBolt" as TransactionBolt
+control kilda.flow
```
Comment (surabujin) [inline docs/design/solutions/hub-and-spoke-tpe/flow_create_install.txt]: move it next to kilda.northbound

File: docs/design/solutions/hub-and-spoke-tpe/flow_create_install.txt [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+NB -> kilda.flow: CommandMessage\nFlowCreateRequest
+kilda.flow -> SplitterBolt: CommandMessage\nFlowCreateRequest
+SplitterBolt -> CrudBolt: CommandMessage\nFlowCreateRequest
+note right of CrudBolt: validate the flow (using FlowValidator)
+note right of CrudBolt: new path (using PathComputer.getPath)
+note right of CrudBolt: build flow pair (state=ALLOCATED)
+note right of CrudBolt: save flow (FlowCache.createFlow)
+note right of CrudBolt: allocate cookie for flow
+note right of CrudBolt: allocate transit vlan (forward)
+note right of CrudBolt: allocate transit vlan (reverse)
```
Comment (surabujin) [inline docs/design/solutions/hub-and-spoke-tpe/flow_create_install.txt]: you forgot meters

File: docs/design/solutions/hub-and-spoke-tpe/flow_create_install.txt [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+note right of CrudBolt: new FlowInfoData
+CrudBolt -> FlowCommandBolt: InfoMessage.FlowInfoData
+CrudBolt -> NorthboundReplyBolt:  FlowInfoData
+alt error
```
Comment (surabujin) [inline docs/design/solutions/hub-and-spoke-tpe/flow_create_install.txt]: `alt` between what and what?

File: docs/design/solutions/hub-and-spoke-tpe/flow_create_install.txt [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+	CrudBolt -> ErrorBolt: ErrorMessage
+	ErrorBolt -> NorthboundReplyBolt: ErrorMessage
+end
+NorthboundReplyBolt -> kilda.northbound: InfoMessage or\n ErrorMessage
+
+loop in forward, reveres
+  loop in segments
+    FlowCommandBolt -> FlowService: create flow
```
Comment (surabujin) [inline docs/design/solutions/hub-and-spoke-tpe/flow_create_install.txt]: This and next operations are not part of "segments" loop.

File: docs/design/solutions/hub-and-spoke-tpe/flow_create_install.txt [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+note right of CrudBolt: new FlowInfoData
+CrudBolt -> FlowCommandBolt: InfoMessage.FlowInfoData
+CrudBolt -> NorthboundReplyBolt:  FlowInfoData
+alt error
+	CrudBolt -> ErrorBolt: ErrorMessage
+	ErrorBolt -> NorthboundReplyBolt: ErrorMessage
+end
+NorthboundReplyBolt -> kilda.northbound: InfoMessage or\n ErrorMessage
+
+loop in forward, reveres
```
Comment (surabujin) [inline docs/design/solutions/hub-and-spoke-tpe/flow_create_install.txt]: Are you going to ignore error from DB(DAO) layer?

PS Don't see where we storing flow in DB.

File: docs/design/solutions/hub-and-spoke-tpe/flow_delete_uninstall.txt [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+control kilda.flow
+participant "FLOW SpeakerBolt" as SpeakerBolt
+
+Client -> NB: /flows DELETE
+NB -> kilda.flow: CommandMessage\FlowDeleteRequest
+kilda.flow -> SplitterBolt: CommandMessage\FlowDeleteRequest
+SplitterBolt -> CrudBolt: CommandMessage\FlowDeleteRequest
+CrudBolt -> FlowCommandBolt: InfoMessage.FlowInfoData
+CrudBolt -> NorthboundReplyBolt:  FlowInfoData
+alt error
```
Comment (surabujin) [inline docs/design/solutions/hub-and-spoke-tpe/flow_delete_uninstall.txt]: ???

File: docs/design/solutions/hub-and-spoke-tpe/flow_delete_uninstall.txt [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+	CrudBolt -> ErrorBolt: ErrorMessage
+	ErrorBolt -> NorthboundReplyBolt: ErrorMessage
+end
+NorthboundReplyBolt -> kilda.northbound: InfoMessage or\n ErrorMessage
+
+loop in forward, reveres
+  loop in segments
+    FlowCommandBolt -> FlowService: delete flow
+    FlowService -> DAO: delete flow
+    note right of DAO: status is In_progress
```
Comment (surabujin) [inline docs/design/solutions/hub-and-spoke-tpe/flow_delete_uninstall.txt]: "in progress"? How are you going to distinguish "in progres" in flow create and "in progress" in flow delete?

File: docs/design/solutions/hub-and-spoke-tpe/flow_delete_uninstall.txt [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+CrudBolt -> NorthboundReplyBolt:  FlowInfoData
+alt error
+	CrudBolt -> ErrorBolt: ErrorMessage
+	ErrorBolt -> NorthboundReplyBolt: ErrorMessage
+end
+NorthboundReplyBolt -> kilda.northbound: InfoMessage or\n ErrorMessage
+
+loop in forward, reveres
+  loop in segments
+    FlowCommandBolt -> FlowService: delete flow
```
Comment (surabujin) [inline docs/design/solutions/hub-and-spoke-tpe/flow_delete_uninstall.txt]: Again - flow and segments are separate objects.

File: docs/design/solutions/hub-and-spoke-tpe/flow_delete_uninstall.txt [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+    FlowService -> DAO: delete flow
+    note right of DAO: status is In_progress
+    FlowCommandBolt -> TransactionBolt: RemoveFlow
+    TransactionBolt -> SPEAKER: RemoveFlow
+    SPEAKER -> SpeakerBolt: RemoveFlow
+    SpeakerBolt -> TransactionBolt: RemoveFlow
+    alt all rules are set
+       TransactionBolt -> FlowService: set flow state to delete
+       FlowService -> DAO: status = DOWN
```
Comment (rtretyak) [inline docs/design/solutions/hub-and-spoke-tpe/flow_delete_uninstall.txt]: Are we going to set flow status to Down or remove the flow from DB?

File: docs/design/solutions/hub-and-spoke-tpe/flow_create_install.txt (hunk 5 above)
Comment (timofei-durakov) [reply docs/design/solutions/hub-and-spoke-tpe/flow_create_install.txt]: see loop above

File: docs/design/solutions/hub-and-spoke-tpe/flow_update_install.txt [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+note right of CrudBolt: build flow pair (state=ALLOCATED)
+note right of CrudBolt: save flow (FlowCache.updateFlow)
+note right of CrudBolt: allocate cookie for flow
+note right of CrudBolt: allocate transit vlan (forward)
+note right of CrudBolt: allocate transit vlan (reverse)
+note right of CrudBolt: allocate meters
+note right of CrudBolt: new FlowInfoData
+loop in forward, reveres
+  opt
+    CrudBolt -> FlowService: update flow
```
Comment (nikitacherevko) [inline docs/design/solutions/hub-and-spoke-tpe/flow_update_install.txt]: here is the same: you allocate new bunch of resources, but didn't release previously used

**Pull request merged**